backend/app/controllers/api/v3/users_controller.rb

```diff
@@ -32,6 +32,394 @@ module V3
+          status, response = check_token_for_permission(Permission.admin_id)
+          render json: response.to_json, status: status.to_sym and return if response[:error]
+        end
+
+        # get the user
+        user = User.find_id(params_user_id)
+        return [ { error: 'Invalid' }, :unauthorized ] unless user
+
+        params_user = params[:user] || {}
+        response, status = user.update_info(params_user)
+        render json: response.to_json, status: status
+      end
+
+      # Update a user's permissions
+      #
+      # <b>API Call:</b>
+      #   POST: /api/v3/users/<id>/update_permissions
+      #   {
+      #     token: <token>
+      #     id: <user_id>,
+      #     user: {
+      #       permission_ids: [ <array_of_permission_ids> ]
+      #     }
+      #   }
+      #
+      # <b>API Return Success:</b>
+      #   STATUS_CODE: 200
+      #   {
+      #     user: {
+      #       id: <user_id>,
+      #       name: <name>,
+      #       login: <login>,
+      #       permission_ids: <permission_ids>
+      #     }
+      #   }
+      #
+      # @!method update_permissions(token, id, user)
+      # @param token [String] a token
+      # @param id [Int] the id of the user
+      # @param user [Hash] the user
+      def update_permissions
+        # Check for admin permissions <or> self
+        status, response = check_token_for_permission(Permission.admin_id)
+        render json: response.to_json, status: status.to_sym and return if response[:error]
+
+        # user who is updating
+        by_user_id = response[:user]["id"]
+
+        # get the user
+        params_user_id = Input.int(params[:id])
+        user = User.find_id(params_user_id)
+        return [ { error: 'Invalid' }, :unauthorized ] unless user
+
+        params_user = params[:user] || {}
+        response, status = user.update_permissions(by_user_id, params_user)
+        render json: response.to_json, status: status
+      end
+
+      # Set agreement to true for <agreement> = /lab|aquarium/.
+      #
+      # <b>API Call:</b>
+      #   POST: /api/v3/users/<id>/agreements/<agreeement>
+      #   {
+      #     token: <token>,
+      #     id: <user_id>,
+      #     agreement: "lab" | "aquarium"
+      #   }
+      #
+      # <b>API Return Success:</b>
+      #   STATUS_CODE: 200
+      #   {
+      #     user: {
+      #       id: <user_id>,
+      #       name: <name>,
+      #       login: <login>,
+      #       permission_ids: <permission_ids>
+      #     }
+      #   }
+      #
+      # @!method agreements(token, id)
+      # @param token [String] a token
+      # @param id [Int] the id of the user
+      def agreements
+        # Check token for any permissions
+        status, response = check_token_for_permission
+        render json: response.to_json, status: status.to_sym and return if response[:error]
+
+        # If not oneself then check token for admin permissions
+        params_user_id = Input.int(params[:id])
```

Comment:
isn't this code is repeated in update user - factor out method

Reply:
I'm not sure I agree? 

The checking for admin permissions is already factored out in this method and in these two calls

          status, response = check_token_for_permission(Permission.admin_id)
          render json: response.to_json, status: status.to_sym and return if response[:error]

All I am doing here is wrapping those two calls if you are trying to update someone else's info (i.e., only admins can update other people's info)

        if response[:user]["id"] != params_user_id
          status, response = check_token_for_permission(Permission.admin_id)
          render json: response.to_json, status: status.to_sym and return if response[:error]
       end

I think wrapping the wrapper (i.e., to check whether you are trying to update someone else's info and if so to check whether you are an admin) is a bit overkill.

I think we have 2 options
option 1- leave as is
option 2- create a new method that checks (a) whether you have a valid token < and > (b) if you have a valid token and if you are trying to update someone else's info then check whether you have admin permissions

This would be wrapping both calls
        # Check token for any permissions
        status, response = check_token_for_permission
        render json: response.to_json, status: status.to_sym and return if response[:error]

        # If not oneself then check token for admin permissions
        params_user_id = Input.int(params[:id])
        if response[:user]["id"] != params_user_id
          status, response = check_token_for_permission(Permission.admin_id)
          render json: response.to_json, status: status.to_sym and return if response[:error]
        end

into a single call, such as "check_token_for_valid_permission_or_admin"

But I still think the "double-check" as currently implemented is clean and just as easy to follow.

Reply:
probably requires some rethinking, but let's leave it

Reply:
OK.